Beyond foundation

EgbertRijke opened the pull request:
In this PR I pulled changes I made in the `foundation` and `foundation-core` folders in the Beyond finite sets PR #623. The goal is to create something mergeable without having to wait for the other unfinished changes in #623. Let's see what monstrosity I created.

EgbertRijke commented on src/foundation/symmetric-operations.lagda.md:
````diff
@@ -1,6 +1,8 @@
 # Symmetric operations
 
 ```agda
+{-# OPTIONS --allow-unsolved-metas #-}
````
Note that this PR has an open hole.

fredrik-bakke replied:
Okay, I will mark it as a draft until the hole is filled.

EgbertRijke commented:
Ready for review

fredrik-bakke commented:
Can we have #706 merged before this one?

fredrik-bakke commented on src/foundation/action-on-equivalences-type-families.lagda.md:
```diff
@@ -0,0 +1,169 @@
+  ( P : subuniverse l1 l2) (B : type-subuniverse P → UU l3)
+  where
+
+  unique-action-on-equivalences-family-of-types-subuniverse :
```
This use of binding words like `on` and `of` in names is not something that is used very much in `foundation` currently, other than in module names. I'm not opposed to using them, in fact, it seems like many names would be more readable with such bindings. However, I find it looks a bit odd next to other definitions that omit them, and since part of the purpose of our naming scheme is to make it easy to guess the names of things, I think it is important that the names are consistent with each other.
Therefore, if we are to start using such bindings, I would suggest that we do a refactor to introduce it everywhere all at once.

fredrik-bakke replied:
Also, do you dislike the term `family` for family of types? It's a lot less verbose, and works more naturally with our naming scheme, I find.

fredrik-bakke replied:
I'm still of the impression that `unique-ap-equiv-subuniverse` could work well as a name for this definition

fredrik-bakke commented on src/foundation/action-on-equivalences-type-families.lagda.md:
Given that this file generalizes/specializes action on equivalences in #706 to subuniverses, perhaps it's more fitting to call this file `action-on-equivalences-subuniverses`?

EgbertRijke commented:
Not sure how this one didn't have any merge conflicts after your big refactor

EgbertRijke commented:
I made all the changes I wanted to make to integrate this well with #706. This one is now ready for review.

EgbertRijke commented:
Oh wait I want to make some more changes:)

EgbertRijke commented:
Now I think I got them all

fredrik-bakke commented:
Great! I will review it this evening :)

fredrik-bakke commented on src/foundation/unordered-pairs.lagda.md:
```diff
@@ -125,21 +129,34 @@ Any two elements `x` and `y` in a type `A` define a standard unordered pair
+  pr1 standard-unordered-pair = Fin-UU-Fin' 2
+  pr2 standard-unordered-pair = element-standard-unordered-pair
+
+  other-element-standard-unordered-pair : Fin 2 → A
+  other-element-standard-unordered-pair (inl (inr star)) = y
+  other-element-standard-unordered-pair (inr star) = x
```
I find this name a little strange. How about `swapped-element-standard-unordered-pair` instead?

EgbertRijke replied:
I found it very natural 😁 you start with one element in a 2-element type/unordered pair, you apply this function and you get the other element.

fredrik-bakke replied:
Alright then 😅

fredrik-bakke reviewed (approved):
Looks very good! Thanks again for cleaning up after me.

fredrik-bakke commented:
One other minor remark is, for completeness, we could also repeat all the computations for action on equivalences of type families over subuniverses, and in the process also create a file for action on equivalences of functions on subuniverses.

fredrik-bakke commented on src/foundation/identity-systems.lagda.md:
```diff
@@ -23,62 +23,76 @@ open import foundation-core.transport-along-identifications
-      (is-contr-AB : is-contr (Σ A B)) →
-      {l : Level} → IND-identity-system l B a b
-    pr1 (Ind-identity-system is-contr-AB P) p x y =
+    is-identity-system-is-torsorial :
```
I guess this is the point of the file on the universal property of identity systems 😅

EgbertRijke commented on src/foundation/identity-systems.lagda.md:
```diff
@@ -23,62 +23,76 @@ open import foundation-core.transport-along-identifications
-      {l : Level} → IND-identity-system l B a b
-    pr1 (Ind-identity-system is-contr-AB P) p x y =
+    is-identity-system-is-torsorial :
+      (H : is-contr (Σ A B)) →
```
I think it does but I haven't looked at that in the past few months.

EgbertRijke replied:
Confirmed. It causes a cyclic dependency.

fredrik-bakke commented on src/foundation/action-on-equivalences-type-families.lagda.md:
```diff
@@ -0,0 +1,187 @@
+  unique-action-equiv-family =
+    is-contr-map-ev-id-equiv (λ Y e → B _ ≃ B Y) id-equiv
+
+  action-equiv-family :
+    {X Y : UU l1} → (X ≃ Y) → B X ≃ B Y
+  action-equiv-family {X} {Y} =
+    equiv-eq ∘ action-equiv-function B
```
oh, I see, my bad!

EgbertRijke commented:
Thanks for the review! I will work on it tonight!

EgbertRijke commented on src/foundation/symmetric-binary-relations.lagda.md:
````diff
@@ -0,0 +1,200 @@
+### Symmetric binary relations
+
+```agda
+symmetric-binary-relation :
+  {l1 : Level} (l2 : Level) → UU l1 → UU (l1 ⊔ lsuc l2)
+symmetric-binary-relation l2 A = symmetric-operation A (UU l2)
````
I capitalized `symmetric-binary-relation` to make it consistent with `Relation`, but I left `symmetric-operation` lower case because I don't think for any other kinds of function types we capitalize their names. I'm actually tempted to make `Relation` and all related names lower case, but not in this PR, and I would like to have your input on this before I would do such a thing.

EgbertRijke commented:
I addressed all your comments. Now it's a question of how much laws you want me to prove for the new actions on equivalences, and whether we want those laws that serve no immediate purpose to hold up the useful refactorings in this PR.

fredrik-bakke commented:
None is fine. It was only a remark
